pkg/controller/middleware/auth.go

```diff
@@ -165,3 +165,90 @@ func (rah *RequireAuthHandler) Handle(next http.Handler) http.Handler {
 		}
 	})
 }
+
+type QuotaKeyIssuanceHandler struct {
+	ctx            context.Context
+	client         *auth.Client
+	db             *database.Database
+	quotaTTL       time.Duration // frequency of quota reset
+	requestsPerTTL int           // count of requests over the ttl window.
+	logger         *zap.SugaredLogger
+}
+
+// QuotaKeyIssuance requires a user is authenticated using firebase auth, that such
+// a user exists in the database, and that said user has not made too many
+// requests
+func QuotaKeyIssuance(ctx context.Context, client *auth.Client, db *database.Database, ttl time.Duration, requestsPerTTL int) *QuotaKeyIssuanceHandler {
+	logger := logging.FromContext(ctx)
+	return &QuotaKeyIssuanceHandler{ctx, client, db, ttl, requestsPerTTL, logger}
+}
+
+func (rah *QuotaKeyIssuanceHandler) Handle(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if err := func() error {
+			// Get the cookie
+			cookie, err := r.Cookie("session")
+			if err != nil {
+				return fmt.Errorf("failed to get cookie: %w", err)
+			}
+
+			// Verify cookie
+			token, err := rah.client.VerifySessionCookie(rah.ctx, cookie.Value)
```

Comment:
hmm, we end up doing this in multiple places - and this also ties it to a user.

Would it be better to have our higher middlewares put the quota identifer into the context and have this enforce quota aginst that.

Then in the DB we can have a quota object that can be tied to an API key or a User

Requires middleware installed int he right order.

Reply:
That's kinda how the rate limiter works - you define the key function based on "something" in the http.Request object.